types/signed_header.go

```diff
@@ -22,7 +25,49 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
 		// and is a programmer bug
 		panic(fmt.Errorf("%T is not of type %T", untrst, untrstH))
 	}
+	err := untrstH.ValidateBasic()
+	if err != nil {
+		return &header.VerifyError{
+			Reason: err,
+		}
+	}
+	sHHash := sH.Header.Hash()
+	if !bytes.Equal(untrstH.LastHeaderHash[:], sHHash) {
+		return &header.VerifyError{
+			Reason: fmt.Errorf("Last header hash %v does not match hash of previous header %v", untrstH.LastHeaderHash[:], sHHash),
+		}
+	}
+	sHLastCommitHash := sH.getLastCommitHash()
+	if !bytes.Equal(untrstH.LastCommitHash[:], sHLastCommitHash) {
+		return &header.VerifyError{
+			Reason: fmt.Errorf("Last commit hash %v does not match hash of previous header %v", untrstH.LastCommitHash[:], sHHash),
+		}
+	}
 	return sH.Header.Verify(&untrstH.Header)
 }
 
 var _ header.Header = &SignedHeader{}
+
+func (sH *SignedHeader) getLastCommitHash() []byte {
+	lastABCICommit := tmtypes.Commit{
+		Height: sH.Height(),
+		Round:  0,
+		BlockID: tmtypes.BlockID{
+			Hash:          tmbytes.HexBytes(sH.Hash()),
+			PartSetHeader: tmtypes.PartSetHeader{},
+		},
+	}
+	for _, sig := range sH.Commit.Signatures {
+		commitSig := tmtypes.CommitSig{
+			BlockIDFlag: tmtypes.BlockIDFlagCommit,
+			Signature:   sig,
+		}
+		lastABCICommit.Signatures = append(lastABCICommit.Signatures, commitSig)
+	}
```

Comment:
Whenever you use `append` you are reallocating memory.  To avoid this, you can pre-allocate the capacity of the slice.
```suggestion
	lastABCICommit := tmtypes.Commit{
		Height: sH.Height(),
		Round:  0,
		BlockID: tmtypes.BlockID{
			Hash:          tmbytes.HexBytes(sH.Hash()),
			PartSetHeader: tmtypes.PartSetHeader{},
		},
		Signatures: make([]tmtypes.CommitSig,0,len(sH.Commit.Signatures)
	}
	for _, sig := range sH.Commit.Signatures {
		commitSig := tmtypes.CommitSig{
			BlockIDFlag: tmtypes.BlockIDFlagCommit,
			Signature:   sig,
		}
		lastABCICommit.Signatures = append(lastABCICommit.Signatures, commitSig)
	}
```

Reply:
I don't think you should append when you've already made allocated. Needs to use `lastABCICommit[i]` most likely.

Reply:
This `getLastCommitHash` is taken from https://github.com/rollkit/rollkit/blob/7bd5df91ed0f48d72c4caac7de069b48b6c1f608/state/executor.go#L478
and
https://github.com/rollkit/rollkit/blob/e4f8bd441b49af44de04b637031520eab437011f/conv/abci/block.go#L119

I tried using the existing method but there's a circular dependency if I try to do that :/

Reply:
That's an interesting problem. If we extracted the header logic into its own file, we could import logic built on other types without having a circular dependency. For now, I think that is fine but we should clean this up again in a follow-up PR.